### ISSUE TITLE
significant-drop-in-scrutinee when false positive

### ISSUE THREAD

**Issue by aviramha** [rust-lang/rust-clippy#8963]
### Summary

Last version of clippy seems to have false positive on the lint
happens in our CI
https://github.com/metalbear-co/mirrord/runs/6770432194?check_suite_focus=true

### Lint Name

significant-drop-in-scrutinee

### Reproducer

I tried this code:

```rust

pub(crate) static SOCKETS: SyncLazy<Mutex<HashSet<Socket>>> =
    SyncLazy::new(|| Mutex::new(HashSet::new()));

fn bind(sockfd: c_int, addr: *const sockaddr, addrlen: socklen_t) -> c_int {
    debug!("bind called sockfd: {:?}", sockfd);
    let mut socket = {
        let mut sockets = SOCKETS.lock().unwrap();
        match sockets.take(&sockfd) {
            Some(socket) if !matches!(socket.state, SocketState::Initialized) => {
                error!("socket is in invalid state for bind {:?}", socket.state);
                return libc::EINVAL;
            }
            Some(socket) => socket,
            None => {
                debug!("bind: no socket found for fd: {}", &sockfd);
                return unsafe { libc::bind(sockfd, addr, addrlen) };
            }
        }
    };
....
```

I saw this happen:

```
   --> mirrord-layer/src/sockets.rs:168:15
    |
168 |         match sockets.take(&sockfd) {
    |               ^^^^^^^^^^^^^^^^^^^^^
    |
    = note: `-D clippy::significant-drop-in-scrutinee` implied by `-D warnings`
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#significant_drop_in_scrutinee
```

### Version

```text
rustc 1.63.0-nightly (50b00252a 2022-06-06) (from rustc 1.63.0-nightly (50b00252a 2022-06-06))
```

**Comment by SpriteOvO** [comment]
Same problem:

```rust
pub(crate) fn default_thread_pool() -> Arc<ThreadPool> {
    static POOL_WEAK: Lazy<Mutex<Weak<ThreadPool>>> = Lazy::new(|| Mutex::new(Weak::new()));

    let mut pool_weak = POOL_WEAK.lock_expect(); // `lock_expect` is a wrapper of `lock().expect("xxx")`

    match pool_weak.upgrade() {
        Some(pool) => pool,
        None => {
            let pool = Arc::new(ThreadPool::builder().build().unwrap());
            *pool_weak = Arc::downgrade(&pool);
            pool
        }
    }
}
```

Warning:

```
warning: temporary with significant drop in match scrutinee
   --> src/thread_pool.rs:155:11
    |
155 |     match pool_weak.upgrade() {
    |           ^^^^^^^^^^^^^^^^^^^
    |
    = note: `#[warn(clippy::significant_drop_in_scrutinee)]` on by default
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#significant_drop_in_scrutinee
```

**Comment by Kixiron** [comment]
The false positives also appear in for loops
[Playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=317855c1762c498b102cd4ff3da6cb51)
```rust
use std::{sync::Mutex, collections::HashSet};

fn main() {
    let nodes: Mutex<HashSet<u8>> = Mutex::new(HashSet::new());
    let mut nodes = nodes.lock().unwrap();
    
    for node in nodes.drain() {
        println!("{node}");
    }
}
```
```rust
warning: temporary with significant drop in for loop
 --> src/main.rs:7:17
  |
7 |     for node in nodes.drain() {
  |                 ^^^^^^^^^^^^^
  |
  = note: `#[warn(clippy::significant_drop_in_scrutinee)]` on by default
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#significant_drop_in_scrutinee

warning: `playground` (bin "playground") generated 1 warning
```
```
$ cargo +nightly -vV
cargo 1.63.0-nightly (4d92f07f3 2022-06-09)
release: 1.63.0-nightly
commit-hash: 4d92f07f34ba7fb7d7f207564942508f46c225d3
commit-date: 2022-06-09
host: x86_64-unknown-linux-gnu
libgit2: 1.4.2 (sys:0.14.2 vendored)
libcurl: 7.83.1-DEV (sys:0.4.55+curl-7.83.1 vendored ssl:OpenSSL/1.1.1n)
os: Ubuntu 20.4.0 [64-bit]
```

**Comment by alice-i-cecile** [comment]
We're hitting a false positive here too: https://github.com/bevyengine/bevy/pull/5160.

**Comment by SpriteOvO** [comment]
For me, this false positive no longer occurs in the latest version.

```
cargo +nightly --version --verbose
cargo 1.65.0-nightly (efd4ca3dc 2022-08-12)
release: 1.65.0-nightly
commit-hash: efd4ca3dc0b89929dc8c5f5c023d25978d76cb61
commit-date: 2022-08-12
host: x86_64-unknown-linux-gnu
libgit2: 1.4.2 (sys:0.14.2 vendored)
libcurl: 7.83.1-DEV (sys:0.4.55+curl-7.83.1 vendored ssl:OpenSSL/1.1.1q)
os: Arch Linux [64-bit]
```
